core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -452,7 +440,10 @@ private[spark] class BlockManager(
         if (level.useMemory) {
           logDebug(s"Getting block $blockId from memory")
           val result = if (asBlockResult) {
-            memoryStore.getValues(blockId).map(new BlockResult(_, DataReadMethod.Memory, info.size))
+            memoryStore.getValues(blockId).map { iter =>
+              val ci = CompletionIterator[Any, Iterator[Any]](iter, releaseLock(blockId))
+              new BlockResult(ci, DataReadMethod.Memory, info.size)
```

Comment:
The difference is that now the programmer needs to explicitly `completionCallback = None`. If the `completionCallback` is specified then you don't need to do the network check. It's better in that today you have zero reminder that you need to release the lock by the end of the task.

Actually an even better way IMO is to have a `LocalBlockResult` and a `RemoteBlockResult` so there's no way the programmer can forget to release the lock.

By the way, I'm not quite done reviewing yet but feel free to address these in a follow-up patch.